File: LEVEL2/delivery.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

queue<int> q;
vector<bool> check;

int solution(int N, vector<vector<int>> road, int K) {
    check.resize(N+1,false);
    vector<int> min_dis(N+1,500001);

    int answer=0;
    q.push(1);
    min_dis[1]=0;
    while(!q.empty())
    {
        int town=q.front();
        if(!check[town] && min_dis[town]<=K)
        {
            check[town]=true;
            answer++;
        }
        for(vector<int> r : road)
        {
            if(r[0]==town && r[1]!=1 && min_dis[r[1]] > r[2]+min_dis[town])
            {
                q.push(r[1]);
                min_dis[r[1]]=r[2]+min_dis[town];
            }
            if(r[1]==town && r[0]!=1 && min_dis[r[0]] > r[2]+min_dis[town])
            {
                q.push(r[0]);
                min_dis[r[0]]=r[2]+min_dis[town];
            }
        }
        q.pop();
    }

    return answer;
}

int main()
{
    int N=5, K=3;
    vector<vector<int>> r=
    {{1,2,1},{2,3,3},{5,2,2},{1,4,2},{5,3,1},{5,4,2}};
    
    cout << solution(N, r, K);
    return 0;
}

File: LEVEL2/delivery.js
function solution(N,road,K){
    let check = new Array(N+1).fill(false);
    let minDistance = new Array(N+1).fill(500001);
    let que = [];
    let answer = 0;
    que.push(1);
    minDistance[1]=0;
    while(que.length !== 0){
        let town = que[0];
        if(!check[town] && minDistance[town]<=K){
            check[town] = true;
            answer++;
        }
        road.forEach(([town1, town2, dis])=>{
            if(town1===town && minDistance[town2]>dis+minDistance[town]){
                que.push(town2);
                minDistance[town2]=dis+minDistance[town];
            }
            if(town2===town && minDistance[town1]> dis+minDistance[town]){
                que.push(town1);
                minDistance[town1]=dis+minDistance[town];
            }
        })
        que.shift();
    }
    return answer;
}
const N = 5;
const road = [[1,2,1],[2,3,3],[5,2,2],[1,4,2],[5,3,1],[5,4,2]];
const K = 3;
console.log(solution(N,road,K));